External signer support - Wallet Box edition

Sjors opened the pull request:
Big picture overview in [this gist](https://gist.github.com/Sjors/29d06728c685e6182828c1ce9b74483d).

This PR lets `bitcoind` call an arbitrary command `-signer=<cmd>`, e.g. a hardware wallet driver,  where it can fetch public keys, ask to display an address, and sign a transaction (using PSBT under the hood).

It's design to work with https://github.com/bitcoin-core/HWI, which supports multiple hardware wallets. Any command with the same arguments and return values will work. It simplifies the manual procedure described [here](https://github.com/bitcoin-core/HWI/blob/master/docs/bitcoin-core-usage.md).

Usage is documented in [doc/external-signer.md](
https://github.com/Sjors/bitcoin/blob/2019/08/hww-box2/doc/external-signer.md), which also describes what protocol a different signer binary should conform to.

Use `--enable-external-signer` to opt in, requires Boost::Process:

```
Options used to compile and link:
  with wallet     = yes
  with gui / qt   = no
  external signer = yes
```

It adds the following RPC methods:
* `enumeratesigners`: asks <cmd> for a list of signers (e.g. devices) and their master key fingerprint
* `signerdisplayaddress <address>`:  asks <cmd> to display an address

It enhances the following RPC methods:
* `createwallet`: takes an additional `external_signer` argument and fetches keys from device
* `send`: automatically sends transaction to device and waits 

Usage TL&DR:
* clone HWI repo somewhere and launch `bitcoind -signer=../HWI/hwi.py`
* check if you can see your hardware device: `bitcoin-cli enumeratesigners`
* create wallet and auto import keys `bitcoin-cli createwallet "hww" true true "" true true true`
* display address on device: `bitcoin-cli signerdisplayaddress ...`
* to spend, use `send` RPC and approve transaction on device

Prerequisites:
- [x] #21127 load wallet flags before everything else
- [x] #21182 remove mostly pointless BOOST_PROCESS macro

Potentially useful followups:
- GUI support: bitcoin-core/gui#4
- bumpfee support
- (automatically) verify (a subset of) keys on the device after import, through message signing

Sjors committed:
test: framework: add skip_if_no_external_signer

Sjors committed:
rpc: add external signer RPC files

Sjors committed:
rpc: signer: add enumeratesigners to list external signers

Sjors committed:
wallet: add -signer argument for external signer command

Create basic ExternalSigner class with contructor. A Signer(<cmd>)
is added to CWallet on load if -signer=<cmd> is set.

Sjors committed:
test: add external signer test

Includes a mock to mimick the HWI interace.

Sjors committed:
wallet: add external_signer flag

Sjors committed:
rpc: add external_signer option to createwallet

Sjors committed:
test: external_signer wallet flag is immutable

Sjors committed:
rpc: send: support external signer

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2653,6 +2673,8 @@ static UniValue createwallet(const JSONRPCRequest& request)
             {"passphrase", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "Encrypt the wallet with this passphrase."},
             {"avoid_reuse", RPCArg::Type::BOOL, /* default */ "false", "Keep track of coin reuse, and treat dirty and clean coins differently with privacy considerations in mind."},
+            {"descriptors", RPCArg::Type::BOOL, /* default */ "false", "Create a native descriptor wallet. The wallet will use descriptors internally to handle address creation"},
+            {"external_signer", RPCArg::Type::BOOL, /* default */ "false", "Use an external signer such as a hardware wallet. Requires -signer to be configured. Wallet creation will fail if keys cannot be fetched. Requires disable_private_keys and descriptors set to true."},
```
This list-of-bools thing seems like a terrible idea...

Sjors replied:
Indeed, I was complaining about that when `blank` was added :-)

khokho1986 replied:
`3AhLfoVCK1UcA8VjCDu6r1ZQTLEdVq3LPc`

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4132,6 +4211,181 @@ UniValue walletcreatefundedpsbt(const JSONRPCRequest& request)
+
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
+        const std::string command = gArgs.GetArg("-signer", DEFAULT_EXTERNAL_SIGNER);
+        if (command == "") throw std::runtime_error(std::string(__func__) + ": restart bitcoind with -signer=<cmd>");
```
What if there are multiple wallets, with different signers?

IMO `-signer` needs to be replaced with either a wallet-stored path, or a path provided when the wallet is loaded...

Sjors replied:
Storing `signer` in the wallet makes sense, but I also think it can wait for a later PR.

khokho1986 replied:
`3AhLfoVCK1UcA8VjCDu6r1ZQTLEdVq3LPc`

khokho1986 replied:
`3AhLfoVCK1UcA8VjCDu6r1ZQTLEdVq3LPc`

Sjors commented:
@Relaxo143 not a chance; this is still work in progress and there's several pull requests that need to be reviewed and merged first. There's also a feature freeze on 0.19.

Sjors commented:
I dropped the dependency on my new `send` RPC proposal #16378, in favor of just tweaking `sendmany` and `sendtoaddress`. It's less powerful, but should reduce the review burden once native descriptors are merged.

Sjors committed:
configure: add --enable-external-signer

This option replaces --with-boost-process

This prepares external signer support to be disabled by default.
It adds a configure option to enable this feature and to check
if Boost::Process is present.

This also exposes ENABLE_EXTERNAL_SIGNER to the test suite via test/config.ini

Sjors committed:
wallet: add GetExternalSigner()

Sjors committed:
wallet: ExternalSigner: add GetDescriptors method

Sjors committed:
doc: add external-signer.md

luke-jr commented on build-aux/m4/ax_boost_process.m4:
```diff
@@ -0,0 +1,121 @@
+#
+#   This macro calls:
+#
+#     AC_SUBST(BOOST_PROCESS_LIB)
```
I don't see where we are linking this

luke-jr commented on configure.ac:
```diff
@@ -1075,6 +1081,7 @@ AX_BOOST_SYSTEM
 AX_BOOST_FILESYSTEM
 AX_BOOST_THREAD
 AX_BOOST_CHRONO
+AX_BOOST_PROCESS
```
This seems like it will pull in Boost.Process (and our - in this case dead - wrapper code using it) even with `--disable-external-signer`.

Maybe not a problem since we plan to use it for other things too?

Sjors committed:
rpc: signerdisplayaddress

Sjors committed:
wallet: add ExternalSignerScriptPubKeyMan

Sjors commented:
Rebase and big update:
* I dropped #15590 (`GetAddressType() & IsSegWit()`) in favor of `GetOutputType()`
* this PR introduces `ExternalSignerScriptPubKeyMan` as a subclass of `DescriptorScriptPubKeyMan`

Sjors commented:
Rebased on the latest #15764.

Sjors commented:
I'm marking this ready for review, even though 372eef4 is a placeholder commit (to be fixed in other pull requests). There's still rough edges, but it does work.

fjahr commented:
Concept ACK

Sjors commented:
Moving to the new GUI repo! bitcoin-core/gui#4

Sjors commented:
Rebased and dropped 3c65029b2cd7dc4874055cc55424bb7512071d2b since it's unused.

Sjors commented:
Rebased on top of the ultimate send RPC #16378; it now uses `send` instead of `sendtoaddress`.

Sjors commented:
Something is causing Travis on ARM64 to bail out right before the functional tests. I already tried a restart.

Sjors commented:
Rebased on Sqlite support #18916

kiminuo commented on src/wallet/externalsigner.h:
```diff
@@ -0,0 +1,70 @@
+    using std::runtime_error::runtime_error;
+};
+
+//! Enables interaction with an external signing device or service, such as a
```
```suggestion
//! Enables interaction with an external signing device or service, such as
```

kiminuo commented on src/wallet/rpcsigner.h:
```diff
@@ -0,0 +1,25 @@
+
+Span<const CRPCCommand> GetSignerRPCCommands();
+
+#endif // BOOST_VERSION
```
```suggestion
#endif // ENABLE_EXTERNAL_SIGNER
```

promag commented:
Nice work, builds on macos and playing around, will try gui branch too.

jonasjoni111 commented:
--enable-debug

fjahr reviewed (commented):
Leaving a first set of questions and comments from my first look at the code. Will do some manual testing soon.

Sjors commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -48,3 +48,18 @@ ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
+
+bool ExternalSignerScriptPubKeyMan::DisplayAddress(const CScript scriptPubKey, const ExternalSigner &signer) const
+{
+#ifdef ENABLE_EXTERNAL_SIGNER
```
I don't think it compiles otherwise, because this calls an `ExternalSigner` function which needs boost::process

fjahr replied:
Oh, that makes sense and I didn't think about that. Might be helpful to note that somewhere in a comment.

felipsoarez reviewed (commented):
utACK

laanwj commented:
There was a merge conflict in `configure.ac` with the NAT-PMP introduction. Rebased version: https://github.com/laanwj/bitcoin/tree/tmp

laanwj commented on src/Makefile.am:
```diff
@@ -250,10 +250,13 @@ BITCOIN_CORE_H = \
   wallet/db.h \
   wallet/dump.h \
+  wallet/externalsigner.h \
+  wallet/external_signer_scriptpubkeyman.h \
```
nit: I'd prefer these to be consistent in naming; either
```
  wallet/externalsigner.h \
  wallet/externalsigner_scriptpubkeyman.h \
```
or
```
  wallet/external_signer.h \
  wallet/external_signer_scriptpubkeyman.h \
```
(same for the associated `.cpp`, of course)

Sjors replied:
I'll go for the second option, because it's more readable.

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -4415,40 +4465,82 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
+        auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new ExternalSignerScriptPubKeyMan(*this, desc));
+        m_spk_managers[id] = std::move(spk_manager);
+#else
+        throw std::runtime_error(std::string(__func__) + ": Wallets with external signers require Boost::System library.");
```
Wouldn't it be better to mention `--enable-external-signer` configure flag in this error message instead of boost (which is just an implementation detail)?

laanwj commented on src/wallet/wallet.h:
```diff
@@ -829,6 +833,12 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
+#ifdef ENABLE_EXTERNAL_SIGNER
+    ExternalSigner GetExternalSigner() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+#endif
+    /* Display address on an external signer. Returns false if external signer support is not compiled */
```
Please use a doxygen compatible comment
```c++
/** Display address on an external signer. Returns false if external signer support is not compiled */
```

laanwj commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -0,0 +1,104 @@
+    return true;
+}
+
+ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
```
Do you really want to construct a new instance every time and return by value here?
This function is called reasonably often (e.g. when signing, when displaying addresses).
It seems that these could be cached, and in that case a pointer (e.g. `std::shared_ptr`) would make more sense.
But not sure.

Sjors replied:
My guess would be that this is negligible compared to the slow USB communication and sluggish hardware wallets in general. I might leave that for someone else to improve.

laanwj commented on src/wallet/externalsigner.h:
```diff
@@ -0,0 +1,70 @@
+    //! Display address on the device. Calls `<command> displayaddress --desc <descriptor>`.
+    //! @param[in] descriptor Descriptor specifying which address to display.
+    //!            Must include a public key or xpub, as well as key origin.
+    UniValue displayAddress(const std::string& descriptor) const;
```
Please start the method names with a capital letter here like in other classes.

laanwj commented on src/wallet/externalsigner.h:
```diff
@@ -0,0 +1,70 @@
+    std::string m_fingerprint;
+
+    //! Bitcoin mainnet or testnet
+    bool m_mainnet;
```
Is a bool enough here? E.g. what if someone wants to use external signers on signet, regtest? Or are those all "testnet" in this description?

Sjors replied:
I should update the comment, but yes: signet, testnet and regtest use the same address format, which is the only thing that seems to matter when using hardware wallets (at the least the devices I've worked with). They don't even have their own BIP44/49/81 coin types.

Sjors replied:
I overlooked that regtest has a different prefix. Maybe an enum is a good idea...
https://github.com/satoshilabs/slips/blob/master/slip-0173.md#registered-human-readable-parts

Sjors commented:
@laanwj thanks for the review! I was waiting for another PR (https://github.com/bitcoin-core/gui/pull/171) to land which I expect to require a rebase, but I might do it sooner.

hebasto commented:
@Sjors 
> @laanwj @fjahr thanks for the reviews! I was waiting for another PR ([bitcoin-core/gui#171](https://github.com/bitcoin-core/gui/pull/171)) to land which I expect to require a rebase, but I might do it sooner.

https://github.com/bitcoin-core/gui/pull/171 just merged :)

Sjors commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -517,8 +517,6 @@ class LegacySigningProvider : public SigningProvider
 class DescriptorScriptPubKeyMan : public ScriptPubKeyMan
 {
 private:
-    WalletDescriptor m_wallet_descriptor GUARDED_BY(cs_desc_man);
```
It's not removed: I moved it to `protected`.

Sjors commented on src/wallet/rpcsigner.cpp:
```diff
@@ -58,6 +59,41 @@ static RPCHelpMan enumeratesigners()
+            std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+            CWallet* const pwallet = wallet.get();
+
+            if (!wallet) return NullUniValue;
```
Oops

Sjors commented on configure.ac:
```diff
@@ -1320,8 +1333,8 @@ AX_BOOST_SYSTEM
 
-dnl Opt-in to boost-process
-AS_IF([ test x$with_boost_process != x ], [ AX_BOOST_PROCESS ], [ ax_cv_boost_process=no ] )
+dnl Opt-in to boost-process, unless external signer support is requested
```
In this PR you have to use `--with-boost-process` in order to enable external signer support. So it's opt-in. But in the GUI followup that's flipped _if_ you have GUI enabled.

Sjors commented on test/functional/wallet_signer.py:
```diff
@@ -0,0 +1,53 @@
+
+    def run_test(self):
+        self.log.info('-signer=%s' % self.mock_signer_path())
+        assert_equal(self.nodes[0].getbalance(), 1250)
```
I dropped the balance sanity check and demoted the signer path to `debug`.

Sjors commented:
Rebased and addressed most feedback.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4548,7 +4558,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "bumpfee",                          &bumpfee,                       {"txid", "options"} },
     { "wallet",             "psbtbumpfee",                      &psbtbumpfee,                   {"txid", "options"} },
-    { "wallet",             "createwallet",                     &createwallet,                  {"wallet_name", "disable_private_keys", "blank", "passphrase", "avoid_reuse", "descriptors", "load_on_startup"} },
+    { "wallet",             "createwallet",                     &createwallet,                  {"wallet_name", "disable_private_keys", "blank", "passphrase", "avoid_reuse", "descriptors", "load_on_startup", "external_signer"} },
```
can remove this diff, and then rebase

Sjors commented:
Rebased after #20012

achow101 commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -63,3 +63,42 @@ bool ExternalSignerScriptPubKeyMan::DisplayAddress(const CScript scriptPubKey, c
+    const std::string command = gArgs.GetArg("-signer", ""); // DEFAULT_EXTERNAL_SIGNER);
+    if (command == "") throw std::runtime_error(std::string(__func__) + ": restart bitcoind with -signer=<cmd>");
+
+    std::string chain = gArgs.GetChainName();
+    const bool mainnet = chain == CBaseChainParams::MAIN;
+    std::vector<ExternalSigner> signers;
+    ExternalSigner::Enumerate(command, signers, mainnet);
+    if (signers.empty()) return TransactionError::EXTERNAL_SIGNER_NOT_FOUND;
+    // TODO: add fingerprint argument in case of multiple signers
+    ExternalSigner signer = signers[0];
```
In 57b26f44b9109eb0c366adeb54bf5ebfc816b45e "rpc: send: support external signer"

Perhaps use `GetExternalSigner()` instead of duplicate code?

achow101 commented on test/functional/mocks/signer.py:
```diff
@@ -52,8 +52,25 @@ def displayaddress(args):
+    if args.fingerprint != "00000001":
+        return sys.stdout.write(json.dumps({"error": "Unexpected fingerprint", "fingerprint": args.fingerprint}))
+
+    f = open(os.path.join(os.getcwd(), "mock_psbt"), "r")
+    mock_psbt = f.read()
+    f.close()
```
In 57b26f44b9109eb0c366adeb54bf5ebfc816b45e "rpc: send: support external signer"

Use `with`

```suggestion
    with open(os.path.join(os.getcwd(), "mock_psbt"), "r") as f:
         mock_psbt = f.read()
```

achow101 commented on test/functional/wallet_signer.py:
```diff
@@ -134,5 +135,85 @@ def run_test(self):
+
+        assert(hww.testmempoolaccept([mock_tx])[0]["allowed"])
+
+        f = open(os.path.join(self.nodes[1].cwd, "mock_psbt"), "w")
+        f.write(mock_psbt_signed["psbt"])
+        f.close()
```
In 57b26f44b9109eb0c366adeb54bf5ebfc816b45e "rpc: send: support external signer"

Use `with`

```suggestion
        with open(os.path.join(os.getcwd(), "mock_psbt"), "w") as f:
             f.write(mock_psbt_signed["psbt"])
```

achow101 commented on test/functional/wallet_signer.py:
```diff
@@ -0,0 +1,52 @@
+        self.skip_if_no_external_signer()
+
+    def set_mock_result(self, node, res):
+        f = open(os.path.join(node.cwd, "mock_result"), "w", encoding="utf8")
+        f.write(res)
+        f.close()
```
In 2926553d26e2f15e7986a73f6ee678112ab8ab7f "test: add external signer test"

Use `with`

```suggestion
        with open(os.path.join(node.cwd, "mock_result"), "w", encoding="utf8") as f:
            f.write(res)
```

achow101 commented on test/functional/wallet_signer.py:
```diff
@@ -0,0 +1,52 @@
+        self.extra_args = [
+            [],
+            ['-signer=%s' % self.mock_signer_path()],
+            ['-signer=%s' % "fake.py"],
```
In 2926553d26e2f15e7986a73f6ee678112ab8ab7f "test: add external signer test"

Use f-strings

```suggestion
            [f"-signer={self.mock_signer_path()}"],
            ["-signer=fake.py"],
```

Sjors replied:
Ah, we upgraded to Python 3.6 "recently" :-)

laanwj replied:
It's funny but also a bit shameful that this has taken so long that the drift of linters and python versions is starting to eat into it.

Sjors replied:
Oh I've been plowing through rebases while descriptor wallets were being built. This is nothing :-)

Sjors commented on build_msvc/bitcoin_config.h:
```diff
@@ -53,6 +53,9 @@
 #define HAVE_BOOST_PROCESS /**/
 
+/* define if external signer support is enabled (requires Boost::Process) */
+#define ENABLE_EXTERNAL_SIGNER 1 /**/
```
Just defining or setting it to 1 always confused me. As does MSVC. But it sounds reasonable... @sipsorcery?

achow101 replied:
I think it's supposed to be either the default or a fixed setting indicating what msvc should be building. Either way, having `HAVE_BOOST_PROCESS` not defined above and still defining `ENABLE_EXTERNAL_SIGNER` is a conflict because we can't have external signer support without boost process.

sipsorcery replied:
I think msvc is the same as the other compilers. If you want to set a pre-processor define to act like a boolean flag there's no need to set a value on it. `#define ENABLE_EXTERNAL_SIGNER` is all you need if you're only checking for its existence with `#ifdef ENABLE_EXTERNAL_SIGNER`.

Sjors replied:
Ok, I'll use that.

Sjors commented on src/wallet/externalsigner.cpp:
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) 2018-2021 The Bitcoin Core developers
```
Good question. It's been under development and publicly visible since 2018 though.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -4428,8 +4429,17 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
 {
-    auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new DescriptorScriptPubKeyMan(*this, desc));
-    m_spk_managers[id] = std::move(spk_manager);
+    if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
```
I made a separate PR for this: #21127

Sjors commented on test/functional/rpc_help.py:
```diff
@@ -105,10 +105,13 @@ def test_categories(self):
             components.append('Wallet')
 
+        if self.is_external_signer_compiled():
+            components.append('Signer')
```
But then I also have to add a line `components.append('Util')`

Sjors commented:
Rebased and addressed @achow101's feedback. It now calls HWI with `--regtest` (unsupported), `--testnet` or `--signet` instead of using `--testnet` for all non-mainnet networks.

laanwj commented:
This has been open for a long time. Hardware wallet support seems quite important. It would be nice if it got some more review and/or ACKs so that we can move forward with it.

laanwj commented:
Linting errors:
```
Python's open(...) seems to be used to open text files without explicitly
specifying encoding="utf8":

test/functional/mocks/signer.py:        with open(mock_result_path, "r") as f:
test/functional/wallet_signer.py:        with open(os.path.join(node.cwd, "mock_result"), "w") as f:
^---- failure generated from test/lint/lint-python-utf8-encoding.sh
```

Sjors commented:
Rebased, appeased linter and switched to using the new HWI `--chain` argument.

gruve-p commented:
utACK https://github.com/bitcoin/bitcoin/pull/16546/commits/e0ce81b4a5d8b8ae617de211747aac5fcad48570

fanquake reviewed (changes requested):
I've only looked at the build system commit. I don't understand the changes here. There should be a single configure option `enable/disable-external-signer`, which defaults to off. `with-boost-process` should be removed entirely. Boost process checking should only occur with  `enable-external-signer`, and if it's not available, configure should fail. We don't need all the duplicate checking for if the wallet is enabled for setting defaults etc.

Sjors commented:
@fanquake boost process is used by `RunCommandParseJSON` (introduced in #15382). This can be used by any feature that involves calling an external command. Hence the separation. However, if you prefer I can still drop `--with-boost-process`. It just means that if we add another feature that uses this, that feature will require a separate configure flag. That's probably fine.

During the development, when switching between branches all the time,  I find `--with-boost-process` convenient, because it works on this branch as well as master. But that's not a strong argument for keeping it.

fanquake commented:
> It just means that if we add another feature that uses this, that feature will require a separate configure flag. That's probably fine.

That's exactly what should happen. Having a `--with-boost-process` configure flag just to be a catch-all for "turn on anything that uses boost process" is not something we want. Significant new features should have their own configure flags.

>  I find --with-boost-process convenient, because it works on this branch as well as master. But that's not a strong argument for keeping it.

No it isn't.

The Boost Process macro itself is mostly just confusing. Boost Process is header only, so having a macro that run a bunch of checks for a library to link against is pointless? I don't really understand why it was written like that in the first place, and am planning on removing it.

meshcollider reviewed (commented):
Great PR Sjors, really really appreciate your upkeep and the clean,easy-to-review commits.

Review in progress...

- [ ] cad9cb16b6e13cec4110e6727986be0f067e2cd0 configure: add --enable-external-signer
- [X] e91c439d4b1afae1a2d595e04e90e4f18e8a7298 msvc: define ENABLE_EXTERNAL_SIGNER
- [X] 48625eca5bbf80a8aed15f185c475e91b6669d09 doc: include Doxygen comments for ENABLE_EXTERNAL_SIGNER
- [X] f7961b77f1bfc10e0b147c08192c39294a09c523 test: framework: add skip_if_no_external_signer
- [X] 1b34c1d86f467c32a6139bbaeed169e6617911c2 wallet: add -signer argument for external signer command
- [X] 790e192fc077c7490089ab9bf63c1fa3dbb19e26 test: add external signer test
- [X] 24bfbdcf165361a9c3a6aa8ed452ba409df11c16 wallet: add external_signer flag
- [X] 4106d7e45f7ab5a48a0b57d64b767306dfd51934 wallet: add ExternalSignerScriptPubKeyMan
- [X] 1ac8ee857089894223597ab5bf350b117301dc77 rpc: add external signer RPC files
- [X] 6d7d5fe4adb8a6f659de7faed1fde7f5c5c3bbf2 rpc: signer: add enumeratesigners to list external signers
- [X] a4851d7b64f5dc99c4a19ee6555d6629af689d3e rpc: add external_signer option to createwallet
- [X] 87a328215c864965867eb1f8a0712abff88f3b89 test: external_signer wallet flag is immutable
- [X] c73376b9824c79aec5166fa0cb5c49f4ee1fb2bb wallet: add GetExternalSigner()
- [x] 345d94e1c346d7f3cee42091e82ed008ae651887 wallet: fetch keys from external signer upon creation
- [x] 2f47ee89ef46289e22cb91c8a9c26c561aa94a24 rpc: signerdisplayaddress
- [x] fb3559c9d167d7d1128ca1db5f604f5292c0bb30 rpc: send: support external signer
- [X] af6c0c39dbf7f7023aaf9da04f1c7c0ad2428059 doc: add external-signer.md
- [X] b27ec70b1a04f3ece972721d54c2e1bc1ce4862a move-only: add underscore to externalsigner.h

meshcollider commented on src/wallet/wallet.cpp:
```diff
@@ -4487,32 +4488,65 @@ void CWallet::SetupDescriptorScriptPubKeyMans()
+        int account = 0;
+        UniValue signer_res = signer.GetDescriptors(account);
+
+        if (!signer_res.isObject()) throw std::runtime_error(std::string(__func__) + ": Unexpect result");
```
nit: `Unexpect` -> `Unexpected`

meshcollider commented on src/wallet/wallet.cpp:
```diff
@@ -4487,32 +4488,65 @@ void CWallet::SetupDescriptorScriptPubKeyMans()
+        if (!signer_res.isObject()) throw std::runtime_error(std::string(__func__) + ": Unexpect result");
+        for (bool internal : {false, true}) {
+            const UniValue& descriptor_vals = find_value(signer_res, internal ? "internal" : "receive");
+            if (!descriptor_vals.isArray()) throw std::runtime_error(std::string(__func__) + ": Unexpect result");
```
same here

meshcollider commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -61,3 +61,31 @@ bool ExternalSignerScriptPubKeyMan::DisplayAddress(const CScript scriptPubKey, c
+    if (complete) return TransactionError::OK;
+
+    std::string strFailReason;
+    if( !GetExternalSigner().SignTransaction(psbt, strFailReason)) {
```
style nit: spacing after `if(`

meshcollider commented on src/wallet/externalsigner.cpp:
```diff
@@ -74,4 +78,51 @@ UniValue ExternalSigner::GetDescriptors(int account)
+    CDataStream ssTx(SER_NETWORK, PROTOCOL_VERSION);
+    ssTx << psbtx;
+
+    // Check if signer fingerpint matches any input master key fingerprint
```
nit: `fingerpint` -> `fingerprint`

meshcollider reviewed (commented):
Code review ACK modulo my inexperience with build system stuff + the rebase on fanquake's PR

Sjors commented:
Rebased and addressed nits.

meshcollider commented:
Will do some testing

achow101 commented:
ACK 85da1ad0442b243e10e7f671e14812e70e74f624

laanwj commented on configure.ac:
```diff
@@ -1393,16 +1394,21 @@ fi
+ AC_DEFINE([HAVE_BOOST_PROCESS],,[define if Boost::Process is available])
+ AC_DEFINE([ENABLE_EXTERNAL_SIGNER],,[define if external signer support is enabled])
+ ],
+ [ AC_MSG_RESULT(no)]
```
I think this is due to a merge conflict: please keep the error message here. Not finding `boost::process` when `--enable-external-signer` is fatal.

Sjors replied:
Oops, good catch, that indeed fell through the rebase cracks.

laanwj commented on src/wallet/external_signer.cpp:
```diff
@@ -0,0 +1,129 @@
+
+const std::string ExternalSigner::NetworkArg() const
+{
+    if (m_chain == CBaseChainParams::MAIN) {
```
I would prefer to introduce a `CBaseChainParams.externalSignerChainName()` instead of an elaborate `if()` here, for this to keep the chain definition data-driven in one place.

Sjors replied:
Alternatively, also out of scope, we could turn `CBaseChainParams` into an enum, so it's easier to use switch statements, and less likely to be forgotten if we add another network type.

laanwj replied:
That'd be another option, but I prefer  to have all this information in one place. It was the point of having chainparams structures in the first place.

laanwj replied:
But good point that they are strings. They also match exactly here:
```
const std::string CBaseChainParams::MAIN = "main";
const std::string CBaseChainParams::TESTNET = "test";
const std::string CBaseChainParams::SIGNET = "signet";
const std::string CBaseChainParams::REGTEST = "regtest";
```
So wouldn't `"" --chain " + m_chain` achieve exactly the same?

Sjors replied:
Good point...

laanwj commented:
Code review ACK 96f991a11a871502d1a39f684994c1e2664c3af7
Verified that the only changes since 85da1ad0442b243e10e7f671e14812e70e74f624 are the `configure.ac` change, and simpllyfing `ExternalSigner::NetworkArg()` (removing a include for `strencodings.h` along the way).

achow101 commented:
re-ACK 96f991a11a871502d1a39f684994c1e2664c3af7

fanquake commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
+#ifdef ENABLE_EXTERNAL_SIGNER
```
From what I can see, `GetExternalSigner()` is only called from code that is inside `ENABLE_EXTERNAL_SIGNER` #idfefs. So why can't we wrap this whole function in an #ifdef, and drop the need for the `throw()`? I think the same goes for most of this file, and some other functions in this PR. It seems weird that when external signing is not enabled, we'd still be compiling functions that aren't called from anywhere, and who's purpose is to throw a runtime error to tell you about external signing and it's need of Boost Process.

Sjors replied:
If turns out I can put the entire `ExternalSignerScriptPubKeyMan` inside `#ifdef`. The most important `throw()`s are already in `wallet.cpp`.

fanquake reviewed (commented):
In 903f277f982ed390dc68114c9c1eb70b4a59bb19 is where we should be able to drop `HAVE_BOOST_PROCESS` entirely. I realise that boost process _could_ potentially be used for other things in future, but currently, it's not, and as this is implemented, there's a bit of an awkward split between `HAVE_BOOST_PROCESS` & `ENABLE_EXTERNAL_SIGNER` for code which, as I understand it, is all essentially the same feature. So I think for now we should just replace the few usages of `HAVE_BOOST_PROCESS` with `ENABLE_EXTERNAL_SIGNER`.

Sjors commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -0,0 +1,91 @@
+
+ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
+#ifdef ENABLE_EXTERNAL_SIGNER
+    const std::string command = gArgs.GetArg("-signer", ""); // DEFAULT_EXTERNAL_SIGNER);
```
Not sure what was going on there, but I put `DEFAULT_EXTERNAL_SIGNER` back.

Sjors replied:
Now I remember, I didn't want a circular inclusion. I just switched to `""`.

Sjors commented:
Rebased onto CI fix. @fanquake I replaced the remaining instances of `HAVE_BOOST_PROCESS` with `ENABLE_EXTERNAL_SIGNER`. In the process I squashed the MSVC and doc commit, since they're now just a simple rename.

laanwj commented:
Please squash the `move-only: add underscore to externalsigner.h` commit as well. It's introduced in this PR, so let's introduce it with the eventual name directly.

Sjors commented:
That may be a rather big rebase hell though, because the commits make incremental changes to these files so I'd have to divide this commit in 10 pieces, squash them in the right place and then deal with rebase conflicts. Unless there's an intelligent Git incantation to rename a file across all commits?

ryanofsky commented:
> Unless there's an intelligent Git incantation to rename a file across all commits?

Something like this will rename the one file. You can extend it to rename other files or do more replacements across commits:

```bash
end=f1824e7af7a8d53f9be50a6c6354511190dcb774
start=b9a262ea0a86f498959e81113522ce26299f1984

git checkout $end
git filter-branch --tree-filter '
  git mv src/wallet/externalsigner.h src/wallet/external_signer.h
  git grep -l externalsigner | xargs sed -i s:wallet/externalsigner.h:wallet/external_signer.h:
' --force $start^..HEAD
```

laanwj commented:
Also git tends to be smart enough to take into account renamed files when rebasing. If you go to the first commit with `git rebase -i`, amend it for editing, `git mv src/wallet/externalsigner.h src/wallet/external_signer.h`, then `git rebase --continue`, that should not cause any merge conflicts.

Sjors commented:
@ryanofsky's thanks! That worked (with some tweaks), and despite warnings from the git, it still compiles and history looks sane. @laanwj I have indeed seen Git behave sanely when it comes to renames, so maybe my worries were not necessary.

laanwj commented:
re-ACK f75e0c1edde39a91cc353b0102638e232def9476

MarcoFalke commented on doc/external-signer.md:
````diff
@@ -0,0 +1,171 @@
+Create a wallet, this automatically imports the public keys:
+
+```sh
+$ bitcoin-cli createwallet "hww" true true "" true true true
````
nit: Would be nice to use named args

MarcoFalke commented on src/wallet/rpcsigner.cpp:
```diff
@@ -0,0 +1,111 @@
+        },
+        RPCExamples{""},
+        [](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue {
+            std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
```
why does this rpc require a wallet?

Sjors replied:
I might get rid of that requirement in a followup.

hebasto commented:
There is a [buggy](https://github.com/bitcoin/bitcoin/pull/21339#issue-583195145) execution path in the `configure` script. Fixed in #21339.

mariosanchezgarcia commented:
Hola yo soy sathosi nakamoto

hebasto commented on test/config.ini.in:
```diff
@@ -23,3 +23,4 @@ RPCAUTH=@abs_top_srcdir@/share/rpcauth/rpcauth.py
 @BUILD_BITCOIND_TRUE@ENABLE_BITCOIND=true
 @ENABLE_FUZZ_TRUE@ENABLE_FUZZ=true
 @ENABLE_ZMQ_TRUE@ENABLE_ZMQ=true
+@ENABLE_EXTERNAL_SIGNER_TRUE@ENABLE_EXTERNAL_SIGNER=true
```
It seems the `@ENABLE_EXTERNAL_SIGNER_TRUE@` substitution does not work for some reasons.

hebasto replied:
nm, forgot to run `./autogen.sh`
